test/sqllogictest/type-promotion.slt

```diff
@@ -277,6 +273,61 @@ EXPLAIN RAW PLAN FOR
 
 EOF
 
+# Check that char casts to text for string functions
+query T multiline
+EXPLAIN RAW PLAN FOR
+    SELECT lower('a'::char);
+----
+%0 =
+| Constant ()
+| Map lower(chartostr(strtochar("a")))
+
+EOF
+
+# Check that varchar promotes to char
+query T multiline
+EXPLAIN RAW PLAN FOR
+    SELECT 'a'::char < 'a'::varchar;
+----
+%0 =
+| Constant ()
+| Map (strtochar("a") < strtochar(strtovarchar("a")))
+
+EOF
+
+
+# Check that char promotes to text
+query T multiline
+EXPLAIN RAW PLAN FOR
+    SELECT 'a'::char < 'a'::text;
+----
+%0 =
+| Constant ()
+| Map (chartostr(strtochar("a")) < "a")
+
+EOF
+
+query T multiline
+EXPLAIN RAW PLAN FOR
+    SELECT 'a'::varchar < 'a'::text;
+----
+%0 =
+| Constant ()
+| Map (varchartostr(strtovarchar("a")) < "a")
```

Comment:
I think this pair of function calls do nothing.

Reply:
What these calls accomplish is admittedly very obscure!

- `strtovarchar` must exist for sake of length-checking the input and properly
  tracking the value's type during planning.
- `varchartostr` feels like it shouldn't need to exist, but we need it in
  planning to express that the value's type has changed from `varchar` to
  `text` in planning.

For more color: before we implemented `sql: invert cast priority for char and
varchar`, we didn't need `varchartostr` this because we just blithely said that
all `varchar` values were synonymous with `text`. However, by reorganizing the
string type category's cast prioritization list, we were no longer able to make
that assertion because we could no longer throw away when a value was `varchar`
vs. `text` because one can be implicitly cast to `char` and one cannot
(`varchar` and `text`, respectively).

In consequence, `varchar` and `text`'s synonymity made it impossible to select a
`char`-based implementation of a function, such as equality. Whenever we saw a
`varchar` value, we would say, "Oh, this is basically just a string" and select
the `text`-based implementation because `text` is the preferred type of the
types' category.

And if we jumped through a number of hoops to more indirectly solve the
aforementioned problem, leaving an expression's `ScalarType` as `varchar` when
we thought it should be `text` would break semantics for operations that
generate homogeneous expressions, such as `coalesce`; we would assume that the
`varchar`-typed expression should be promoted to a `char`, when what should
really happen is that the `char` expression should be promoted to `text`.

I think we've minimized the impact that this will actually make to users,
though, by `varchartostr` an essentially meaningless pass-through function.

Reply:
The impact would be that a particular plan may be worse because an implicit cast would prevent various optimizations from kicking in and certain indexes from being used. There is a separate ticket for that #4171. Once that ticket is fixed hopefully this particular problem will become a non-issue.